examples/llm/src/models/layers/attention.py

```diff
@@ -237,11 +266,34 @@ def __init__(self, cfg: DictConfig, device: Optional[str] = None):
             'using `attn_impl: flash`.')
 
     def forward(self, x, key_padding_mask=None, attn_mask=None):
-        assert key_padding_mask is None
-        return self.mhsa(x,
-                         key_padding_mask=None,
-                         attn_mask=attn_mask,
-                         need_weights=False)
+        if self.attn_qk_ln or self.clip_qkv:
+            qkv = self.Wqkv(x)
+            if self.clip_qkv:
+                qkv = qkv.clamp_(min=-self.clip_qkv, max=self.clip_qkv)
```

Comment:
if using `clamp_` do you need to do the assignment?

Reply:
making valid points!